### PR TITLE
Refactor worker/uniter/relation for interfaces

### PR DESCRIPTION
**Pull request by hmlanigan**
Prep work for mocking uniter/relation tests.

The api/uniter Relation changes were necessary to make a usable Relation interface which could retrieve a RelationUnit.

Interface changes in uniter/relation, flowed over into runner/context to allow for calling NewContextRelation.

### CODE REVIEW

File: api/uniter/relation.go [hunk 1]
```diff
@@ -114,18 +112,16 @@ func (r *Relation) Endpoint() (*Endpoint, error) {
...
 }
 
-// Unit returns a RelationUnit for the supplied unit.
-func (r *Relation) Unit(u *Unit) (*RelationUnit, error) {
-	if u == nil {
-		return nil, fmt.Errorf("unit is nil")
-	}
-	result, err := r.st.relation(r.tag, u.tag)
+// Unit returns a RelationUnit for the supplied unitTag and applicationTag.
+func (r *Relation) Unit(uTag names.UnitTag, aTag names.ApplicationTag) (*RelationUnit, error) {
```
Comment (hmlanigan) [inline api/uniter/relation.go:116]: This should really be Unit(uTag names.UnitTag).  We can get the application tag easily enough via the names pkg.

**Comment by hmlanigan** [comment]
Need to fix the go vet issues.

File: worker/uniter/relation/statetracker.go [hunk 2]
```diff
@@ -100,8 +37,8 @@ type RelationStateTrackerConfig struct {
 
 // relationStateTracker implements RelationStateTracker.
 type relationStateTracker struct {
-	st              *uniter.State
-	unit            *uniter.Unit
+	st              StateTrackerState
```
Comment (howbazaar) [inline worker/uniter/relation/statetracker.go:40]: This feels nicer to read. Thank you.

File: worker/uniter/runner/context/relation.go [hunk 3]
```diff
@@ -15,14 +15,66 @@ import (
 	"github.com/juju/juju/worker/uniter/runner/jujuc"
 )
 
+type Relation interface {
```
Comment (howbazaar) [inline worker/uniter/runner/context/relation.go:18]: Is there a reason we don't have these interfaces in `interface.go`?

**Comment by hmlanigan** [comment]
!!build!!

File: worker/uniter/runner/context/relation.go (hunk 3 above)
Comment (hmlanigan) [reply worker/uniter/runner/context/relation.go:18]: interface.go for this pr is in a different package, relation.   I wasn't focused on interfaces for the context package.  They can be moved together in a next step.

**Comment by hmlanigan** [comment]
Force push to resolve merge conflicts.

**Comment by hmlanigan** [comment]
Ran the mk8s charm qa steps from #11526

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
intermittent fail in "RunMockContextSuite.TestRunActionCancelled"

$$merge$$

File: worker/uniter/relation/interface.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/core/life"
+	"github.com/juju/juju/core/relation"
+	"github.com/juju/juju/core/watcher"
+	"github.com/juju/juju/worker/uniter/hook"
+	"github.com/juju/juju/worker/uniter/remotestate"
+	"github.com/juju/juju/worker/uniter/runner/context"
+)
+
+//go:generate go run github.com/golang/mock/mockgen -package mocks -destination mocks/mock_statetracker.go github.com/juju/juju/worker/uniter/relation RelationStateTracker
```
Comment (manadart) [inline worker/uniter/relation/interface.go:19]: @SimonRichardson and I have been putting these in _package_test.go_ as a convention.

**Pull request merged**